### PR TITLE
Add CraftTweaker Integration

### PR DESCRIPTION
**Pull request by Matryoshika**
Adds CraftTweaker Integration for the HammerCraftingTable:
Supports two methods: addRecipe(output, input...) & removeRecipe(output)
_Changes_ to base repo are loading CraftTweaker during preInit, and turning HammerCraftingManager#REGISTRY public.

### CODE REVIEW

**Comment by Matryoshika** [comment]
Example CraftTweaker Script, with emphasized ingredient array
https://pastebin.com/dLbnNqm1

**Comment by thebrightspark** [comment]
Thanks very much!
I see no issues with this, so I'll merge it and it'll be ready for the next release :)

File: src/main/java/brightspark/sparkshammers/integration/crafttweaker/CraftTweakerIntegration.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package brightspark.sparkshammers.integration.crafttweaker;
+
+import crafttweaker.CraftTweakerAPI;
+import net.minecraftforge.fml.common.Optional;
+
+public class CraftTweakerIntegration {
```
Comment (kindlich) [inline src/main/java/brightspark/sparkshammers/integration/crafttweaker/CraftTweakerIntegration.java:6]: You don't need to manually register the class, you can simply add
[`@ZenRegister`](http://crafttweaker.readthedocs.io/en/latest/#Dev_Area/ZenAnnotations/Annotation_ZenRegister/) to your main class

That way, you don't have to check if the mod is loaded either

File: src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+		}
+		
+	}
+	
+	@ZenMethod
+	static public void addRecipe(IItemStack _output, IItemStack[] _input) {
+		ItemStack output = null;
+		List<ItemStack> input = new ArrayList<ItemStack>();
+		try {
+			output = CraftTweakerMC.getItemStack(_output);
```
Comment (kindlich) [inline src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java:97]: You can use `CrafttweakerMC.getItemStacks(IItemStack... ingredientes)`

File: src/main/java/brightspark/sparkshammers/SparksHammers.java [hunk 3]
```diff
@@ -59,6 +61,8 @@ public void preInit(FMLPreInitializationEvent event)
         //Initialize item, blocks, textures/models and configs here
 
         ConfigurationHandler.init(new File(Reference.CONFIG_DIR, "config.cfg"));
+        if(Loader.isModLoaded("crafttweaker"))
```
Comment (kindlich) [inline src/main/java/brightspark/sparkshammers/SparksHammers.java:64]: You don't need this if you use `@ZenRegister` (see below)

File: src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+		/**
+		 * Adds a recipe.
+		 * Due to {@link net.minecraftforge.common.crafting.CraftingHelper} the ingredients
+		 * need to be parsed individually.
+		 */
+		@Override
+		protected void add() {
+			HammerCraftingManager.REGISTRY.register(new HammerShapedOreRecipe(output, 
+					"ABCDE", "FGHIJ", "KLMN ", 
+					'A', input.get(0),  'B', input.get(1),  'C', input.get(2),  'D', input.get(3), 'E', input.get(4),
```
Comment (kindlich) [inline src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java:52]: Prone to `indexOutOfBound` Exceptions if the list's size is smaller than 14
You could go for something like a `NonNullList<ItemStack>` or similar

File: src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+		
+	}
+	
+	@ZenMethod
+	static public void addRecipe(IItemStack _output, IItemStack[] _input) {
+		ItemStack output = null;
+		List<ItemStack> input = new ArrayList<ItemStack>();
+		try {
+			output = CraftTweakerMC.getItemStack(_output);
+			Arrays.asList(_input).forEach(iitemstack -> input.add(CraftTweakerMC.getItemStack(iitemstack)));
```
Comment (kindlich) [inline src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java:98]: Also, why stream here?

File: src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+		
+		/**
+		 * Only looks up and removes a recipe based on the output
+		 */
+		@Override
+		protected void remove() {
+			HammerShapedOreRecipe wanted = null;
+			try {
+				wanted = HammerCraftingManager.getRecipes().stream().filter(recipe -> recipe.getRecipeOutput().isItemEqual(output)).findFirst().get();
```
Comment (kindlich) [inline src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java:66]: Would only remove the first recipe with that output, you might want do go for a `foreach` or similar

**Review by kindlich — COMMENTED** [review]
Also, you don't need your try/catch blocks for the getItemStack methods.

And:
Normally you do `public static` not `static public`

File: src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java (hunk 6 above)
Comment (Matryoshika) [reply src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java:66]: Due to the registry's setup, it is impossible for more than 1 recipes to exist with equal outputs.
[HammerShapedOreRecipe:43](https://github.com/thebrightspark/SparksHammers/blob/1.12/src/main/java/brightspark/sparkshammers/hammerCrafting/HammerShapedOreRecipe.java#L43)

File: src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java (hunk 5 above)
Comment (Matryoshika) [reply src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java:98]: Personal bias toward lambda-forEach over for-loop. Converting the array to a list for that purpose.

File: src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java (hunk 4 above)
Comment (Matryoshika) [reply src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java:52]: To make it "idiot-proof" the best course of action would likely be to replace the `!list.isEmpty();` to `list.size() == 14;` at [CTHammerCraftingTableRecipes:103](https://github.com/thebrightspark/SparksHammers/blob/1.12/src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java#L103)

File: src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java (hunk 2 above)
Comment (Matryoshika) [reply src/main/java/brightspark/sparkshammers/integration/crafttweaker/CTHammerCraftingTableRecipes.java:97]: Not plausibe as of now due to the null-check issue we discussed in Discord

**Pull request merged**